doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+4. It seems useful to allow the results returned by the hooks to be partial i.e. hooks don't have to clone the result returned by the next (default) hook to override it, instead they only need to return an object with properties that they wish to override. This can save the overhead of excessive clones.
+5. It seems `shortCircuit` is not really necessary if hooks can just choose to not call the next hook?
+
+## Hooks
+
+### `resolve`: from specifier to url
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   conditions?: string[],
+ *   importAttributes?: object[]
+ * }} ModuleResolveContext
+ */
+/**
+ * @typedef {{
+ *   url?: string,
+ *   format?: string
+ * }} ModuleResolveResult
+ */
+/**
+ * @param {string} specifier
+ * @param {ModuleResolveContext} context
+ * @param {(specifier: string, context: ModuleResolveContext) => ModuleResolveResult} nextResolve
+ * @returns {ModuleResolveResult}
+ */
+function resolve(specifier, context, nextResolve) {
+  if (shouldOverride(specifier, context.parentURL)) {
+    const url = customResolve(specifier, context.parentURL);
+    return { url };
+  }
+
+  const resolved = nextResolve(specifier, context);
+  if (resolved.url.endsWith('.zip')) {
+    return { format: 'zip' };
+  }
+  return {};  // no override
+}
+```
+
+Notes:
+
+1. Example use case: yarn pnp ([esm hooks](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-pnp/sources/esm-loader/hooks/resolve.ts), [cjs hooks](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-pnp/sources/loader/applyPatch.ts))
+2. `importAttributes` are only available when the module request is initiated with `import`.
+3. For the CJS loader, `Module._cache` is keyed using file names, and it's used in the wild (usually in the form of `require.cache`) so changing it to key on URL would be breaking. We'll need to figure out another way to map it with an url.
+   1. It might be non-breaking to maintain an additional map on the side for mapping the same url with different searches and hashes to a different module instance, or leave `Module._cache` as an alias map for the first module instance mapped by the URL minus search/hash. If in the same realm, `Module._cache` gets directly accessed from the outside, we can emit a warning about the incompatibility. Down the road this may require making `Module._cache` a proxy to propagate delete requests to the actual URL-based map.
+   2. Changing CJS modules to be mapped with URL will have performance consequences from but might also help with the hot module replacement use case. Note that for parent packages with `exports` or `imports` in their `package.json` the URL conversion is already done (and not even cached) even in the CJS loader.
+
+### `load`: from url to source code
+
+```js
+/**
+ * @typedef {{
+ *   format?: string,
+ *   conditions?: string[],
+ *   importAttributes?: object[]
+ * }} ModuleLoadContext
+ */
+/**
+ * @typedef {{
+*   format?: string,
```

Comment:
`format` is required, if we’re copying the existing `load`.

Reply:
Actually this bit seems problematic if we want to delegate ESM detection to the default load - if hooks don't care about the format but simply want to override the loading (e.g. don't load from disk), they need to detect the format themselves, in the absence of `context.format` and call to `nextLoad()`. But that's a pre-existing problem.

Reply:
With format detection, can we make `format` optional, since we will have a source to perform the check on?

Reply:
Alternatively if we want to be explicit, perhaps `context.detectFormat(source)`?

Reply:
They not only need to know how to detect the format, but also need to know when the detection is needed (for example, if there is a `--no-detect-format`, they need to take care of it). At that point we might as well just make the format optional, and Node.js will decide whether format detection is needed after the entire load chain finishes, not during the load step.

Reply:
I think this is all unrelated. Basically currently there’s no way for hooks to both transform a file and then have Node run detection on it; that’s perhaps something we want to enable somehow, but how to do so is a separate concern that isn’t very related to moving the hooks to be sync. One obvious way that comes to mind is to just expose the `containsModuleSyntax` method as a helper, that hooks can run themselves to take advantage of Node’s built-in detection ability.

But anyway I stand by my comment, that for the moment this is required and there’s a good chance it will remain so.